LT-18376 & LT-18437 liblcm AnyCPU updates

gtryus opened the pull request:
This adds AnyCPU to the projects that need it so that all projects build with PlatformTarget of AnyCPU
When the Platform x86 is chosen, that information is passed to the Model.Core so that the right
unmanaged icu dependencies can be copied.
The dependencies are updated to the latest master branch of libpalaso
The test is fixed which failed
The binary artifacts are included for x86 and x64 unmanaged icu dlls
(10 tests fail on AnyCPU which are related to hunspell)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/liblcm/2)
<!-- Reviewable:end -->

papeh commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[LCM.sln, line 100 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjwaIOpC1_Eyn5MMSY:-KnjwaIOpC1_Eyn5MMSZ:bx4xr1z) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/LCM.sln#L100)):*
> ```
> 		{E5E9DDC7-2855-4D92-AD46-960AC4C46457}.Debug|Any CPU.Build.0 = Debug|Any CPU
> 		{E5E9DDC7-2855-4D92-AD46-960AC4C46457}.Debug|x86.ActiveCfg = Debug|Any CPU
> 		{E5E9DDC7-2855-4D92-AD46-960AC4C46457}.Debug|x86.Build.0 = Debug|Any CPU
> ```

I would appreciate a comment about why some of these `x86`s are mapped to `Any CPU`

---

*[lib/teamcity.dep, line 48 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjvcVaqB_zdYH5u5WM:-KnjvcVaqB_zdYH5u5WN:b-40a68a) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/lib/teamcity.dep#L48)):*
> ```
> RevisionValue=latest.lastSuccessful
> Condition=Windows
> Path=gennorm2.exe => downloads/
> ```

is gennorm2.exe architecture-specific? It is copied to `downloads/x64` below. If it is AnyCPU, only one needs to be copied.

---

*[src/SIL.LCModel.Core/SIL.LCModel.Core.csproj, line 22 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knjx_FVvpS3L-LbAQ8m:-Knjx_FVvpS3L-LbAQ8n:bn03exl) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/src/SIL.LCModel.Core/SIL.LCModel.Core.csproj#L22)):*
> ```XML
>     <DefineConstants>DEBUG;TRACE</DefineConstants>
>     <DebugType>full</DebugType>
>     <PlatformTarget>AnyCPU</PlatformTarget>
> ```

I think I would rather have only `*|AnyCPU` PropertyGroups than have two that do the same thing.

---

*[src/SIL.LCModel.Core/SIL.LCModel.Core.csproj, line 206 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knjy1QLg06mjo6S6ew1:-Knjy1QLg06mjo6S6ew2:bopw5wv) ([raw file](https://github.com/sillsdev/liblcm/blob/23fd53d0b316ca93b95d1ecc822662f1343b291a/src/SIL.LCModel.Core/SIL.LCModel.Core.csproj#L206)):*
> ```XML
>   </ItemGroup>
>   <ItemGroup>
>     <Content Include="..\..\lib\downloads\icudt54.dll" Condition="'$(OS)'=='Windows_NT'">
> ```

Is the Windows_NT condition no longer needed?

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1496 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjznxzISMEOxi6xglI:-KnjznxzISMEOxi6xglJ:b-mxfzmc) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1496)):*
> ```Smalltalk
> 			}
> 
> 			private void AppendAttribute(XmlNode annotationDefn, string beginoffset, string toString)
> ```

`AppendAttribute(XmlNode parent, string attrName, string attrVal)`

Rather than reimplementing XmlUtils.AppendAttribute, why not call it and assert these things before or after the call.

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1498 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knk0ix3BiGBTXLb_ZCI:-Knk0ix3BiGBTXLb_ZCJ:bh6nx0g) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1498)):*
> ```Smalltalk
> 			private void AppendAttribute(XmlNode annotationDefn, string beginoffset, string toString)
> 			{
> 				Debug.Assert(annotationDefn.OwnerDocument != null, "annotationDefn.OwnerDocument != null");
> ```

Is `Assert.NotNull(parent.OwnerDocument)` available here?

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1500 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjzqWfBx_kxtnXspcP:-KnjzqWfBx_kxtnXspcQ:b53z0un) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1500)):*
> ```Smalltalk
> 				Debug.Assert(annotationDefn.OwnerDocument != null, "annotationDefn.OwnerDocument != null");
> 				var xa = annotationDefn.OwnerDocument.CreateAttribute(beginoffset);
> 				xa.InnerText = toString;
> ```

`xa.Value` would be clearer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/liblcm/2)*
<!-- Sent from Reviewable.io -->

gtryus commented:
Review status: 4 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[LCM.sln, line 100 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjwaIOpC1_Eyn5MMSY:-KnkBcAvGkSGqcnVMbTz:b2n1tyo) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/LCM.sln#L100)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

I would appreciate a comment about why some of these `x86`s are mapped to `Any CPU`
</blockquote></details>

I added the comment to the commit message. Is this where you want it? I don't think we normally put comments in the .sln file, right?

---

*[lib/teamcity.dep, line 48 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjvcVaqB_zdYH5u5WM:-KnkEupNTCXEeEH8eiyd:b-sa0xz7) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/lib/teamcity.dep#L48)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

is gennorm2.exe architecture-specific? It is copied to `downloads/x64` below. If it is AnyCPU, only one needs to be copied.
</blockquote></details>

Actually it is architecture specific so I have put logic into the targets file to use an architecture specific version. It may be that both architecture versions perform the same thing but it is written in unmanaged code.

---

*[src/SIL.LCModel.Core/SIL.LCModel.Core.csproj, line 206 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knjy1QLg06mjo6S6ew1:-KnkFGCNQkSwm7EdO0U2:bljvkio) ([raw file](https://github.com/sillsdev/liblcm/blob/23fd53d0b316ca93b95d1ecc822662f1343b291a/src/SIL.LCModel.Core/SIL.LCModel.Core.csproj#L206)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

Is the Windows_NT condition no longer needed?
</blockquote></details>

I have to test on Linux to be sure but I don't think it will actually matter.

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1496 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjznxzISMEOxi6xglI:-KnkFXeA0S0owkKZlLmI:b-e137u7) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1496)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

`AppendAttribute(XmlNode parent, string attrName, string attrVal)`

Rather than reimplementing XmlUtils.AppendAttribute, why not call it and assert these things before or after the call.
</blockquote></details>

Apparently this method was removed in the master branch of libpalaso and I am not sure why.

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1498 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knk0ix3BiGBTXLb_ZCI:-KnkIyZbJNvAi_7BC_MY:b4rvn2j) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1498)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

Is `Assert.NotNull(parent.OwnerDocument)` available here?
</blockquote></details>

To me the Assert implies we are testing it so we think there are cases where it could be null and the Debug.Assert implies it could never be null. But if you really want it changed. I can do it.

---

*[tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs, line 1500 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-KnjzqWfBx_kxtnXspcP:-KnkIaBB6rM57L0IbcJz:b-r7x22y) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/tests/SIL.LCModel.Tests/DomainServices/ParagraphParserTests.cs#L1500)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

`xa.Value` would be clearer
</blockquote></details>

It is probably a question of preference but I changed it your way. I find that with elements, Value doesn't work and InnterText or InnerXml does with Attributes, they work the same. So InnerText works consistently for both elements and attributes which is why I tend to use it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/liblcm/2)*
<!-- Sent from Reviewable.io -->

gtryus commented:
Review status: 4 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/SIL.LCModel.Core/SIL.LCModel.Core.csproj, line 22 at r1](https://reviewable.io:443/reviews/sillsdev/liblcm/2#-Knjx_FVvpS3L-LbAQ8m:-KnkKmvBkFQqD5VVf4HO:bu1zl6k) ([raw file](https://github.com/sillsdev/liblcm/blob/f9642c6f13578708bfe09f451b144ab4be1c2d85/src/SIL.LCModel.Core/SIL.LCModel.Core.csproj#L22)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

I think I would rather have only `*|AnyCPU` PropertyGroups than have two that do the same thing.
</blockquote></details>

Visual Studio doesn't understand the use of * here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/liblcm/2)*
<!-- Sent from Reviewable.io -->
